### DISCUSSION TITLE
Consider updating README.md to use virtual environments

### DISCUSSION THREAD

**Discussion by mvilain** [sherlock-project/sherlock#1060] [Ideas]
The current README instructs users to clone this repo to their systems and install modules required to run sherlock on their systems.

I recommend having users use python's venv virtualize module so their base python3 installation is untouched.

```
diff --git a/README.md b/README.md
index e6f441a..71505eb 100644
--- a/README.md
+++ b/README.md
@@ -39,8 +39,10 @@ $ git clone https://github.com/sherlock-project/sherlock.git

 # change the working directory to sherlock
 $ cd sherlock
+$ python3 -m venv .

-# install the requirements
+# activate virtual environment and install the requirements
+$ source ./bin/activate
 $ python3 -m pip install -r requirements.txt
 ```

**Comment by 00-kat** [comment]
Also, on some Linux distros (I know Arch Linux and OpenSUSE Tumbleweed do this, I think Fedora too), using pip will result in an error saying that you should use your distro's package manager or create a virtual environment.

E.G. on Arch Linux:
```
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try 'pacman -S
    python-xyz', where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Arch-packaged Python package,
    create a virtual environment using 'python -m venv path/to/venv'.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip.
    
    If you wish to install a non-Arch packaged Python application,
    it may be easiest to use 'pipx install xyz', which will manage a
    virtual environment for you. Make sure you have python-pipx
    installed via pacman.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
```

Edit 2: Also, this wouldn't be as straight-forward as that diff, since on Windows it's `.\env\`**`Scripts`**`\activate.{bat,ps1}`, and on Linux you need to choose the right file in `./env/bin/` depending on your shell, and `source` should be `.` according to POSIX, and some shells might not have a script in `./env/bin/`, etc. Not to forget that you can't run `.\env\Scripts\activate.ps1` if you don't enable script execution (which is a pain since there's so many ways a simple "paste this in" could go wrong).

**Reply by maniacs2006** [reply]
hi soo what was the solution that you found , stuck on the same issue.

**Reply by 00-kat** [reply]
Make a virtual environment.  On Windows, in Command Prompt:
```console
$ python -m venv env
$ .\env\Scripts\activate.bat
$ <run command as usual>
```
On UNIX-like systems (i.e. macOS, Linux, BSDs, etc.):
```console
$ python3 -m venv env
$ . env/bin/activate
$ <run command as usual>
```

**Reply by maniacs2006** [reply]
thank you soo much

**Comment by mvilain** [comment]
Abandoned this tool.

**Comment by KerberoSec** [comment]
please solve this error, i am literally cryping to solve this error from last few weeks, this error is comming when i tried to start any kali inux tool, please solve it, /home/kali/.local/lib/python3.12/site-packages/requests/__init__.py:89: RequestsDependencyWarning: urllib3 (2.2.3) or chardet (3.0.4) doesn't match a supported version!
  warnings.warn("urllib3 ({}) or chardet ({}) doesn't match a supported "
usage: sherlock [-h] [--version] [--verbose] [--folderoutput FOLDEROUTPUT] [--output OUTPUT] [--tor] [--unique-tor] [--csv] [--xlsx] [--site SITE_NAME]
                [--proxy PROXY_URL] [--dump-response] [--json JSON_FILE] [--timeout TIMEOUT] [--print-all] [--print-found] [--no-color] [--browse]
                [--local] [--nsfw]
                USERNAMES [USERNAMES ...]
sherlock: error: the following arguments are required: USERNAMES

**Reply by ppfeister** [reply]
Try using the official version from PyPI. We aren't affiliated with the maintainers of the Kali package and bugs relating to those packages should be reported to them https://bugs.kali.org
